### ISSUE TITLE
Syntactic converter

### ISSUE THREAD

**Issue by xeno-by** [scalameta/paradise#84]
Yet uncovered tree nodes (courtesy of @olafurpg):
- [ ] Ctor.Ref.Function
- [ ] Ctor.Ref.Name
- [ ] Ctor.Ref.Project
- [ ] Ctor.Ref.Select
- [x] Defn.Macro (#119)
- [ ] Enumerator.Generator
- [ ] Enumerator.Guard
- [ ] Enumerator.Val
- [ ] Mod.Inline
- [ ] Pat.ExtractInfix
- [x] Pat.Type.And (inapplicable for Scala 2.x)
- [ ] Pat.Type.ApplyInfix
- [ ] Pat.Type.Existential
- [ ] Pat.Type.Function
- [x] Pat.Type.Or (inapplicable for Scala 2.x)
- [ ] Pat.Type.Placeholder
- [ ] Pat.Type.Project
- [ ] Pat.Type.Refine
- [ ] Pat.Type.Tuple
- [ ] Pat.Xml
- [ ] Term.ApplyUnary
- [ ] Term.For
- [ ] Term.ForYield
- [ ] Term.TryWithTerm
- [ ] Term.Update
- [ ] Term.Xml
- [x] Type.And (inapplicable for Scala 2.x)
- [ ] Type.Arg.Repeated
- [x] Type.Or (inapplicable for Scala 2.x)
- [ ] Type.Placeholder

Out of a corpus of ~26k source files (from scala/scala, intellij-scala, scala.js, akka, ...), the converter crashes on (only) ~120 source files, split between these errors

```
ConvertException: unexpected result: expected = Pat, actual = Pat.Type.Apply: 2
ConvertException: PackageDef: 2
ConvertException: unexpected result: expected = Pat.Type, actual = Type.Function: 3
ConvertException: Bind: 5
ConvertException: Ident: 5
ConvertException: ModuleDef: 11
ConvertException: unexpected result: expected = Pat.Type, actual = Type.Project: 12
UnreachableError: this code path should've been unreachable: 17
ConvertException: unexpected error (scroll down the stacktrace to see the cause):: 66
```

**Comment by Dveim** [comment]
Not sure about given categories differentiation -- shouldn't we keep `scala.meta.Trees` / `LogicalTrees` as basis? For example, what `Lambdas` here mean, `Term.Function` ? So, I'd suggest to use leaves  of `Trees` for categories, that sounds more intuitive for me.

Meanwhile, I'd take `imports`.

**Comment by xeno-by** [comment]
@Dveim I think it would be more practical to support the most frequent syntactic patterns first. For instance, even though `Term.Function` is just one AST node, it's much more popular than a bunch of `Pat.Type` nodes. That's why I think we should take the aforementioned Eden test suite and try to make it work, and only then go for supporting an exhaustive list of AST nodes.

**Comment by xeno-by** [comment]
Today I'll be working on members.

**Comment by olafurpg** [comment]
Regarding context bounds, they are desugared like this

```scala
// before
def f[T:Classtag] = ???
// after
def f[T](implicit evidence: Classtag[T]) = ???
```

What means do we have to figure out which syntax was originally used?

**Comment by xeno-by** [comment]
You check the name of the implicit parameter. If it contains a dollar, then you collapse into the context bound.

**Comment by xeno-by** [comment]
My work on modifiers and members got blocked by our infrastructure around converters. After #91 and #98, I'm more or less happy with the infrastructure, but now I have to context switch. Therefore, I'm yielding members and modifiers to @olafurpg.

**Comment by Dveim** [comment]
I'd take `Lambdas`.

**Comment by olafurpg** [comment]
Cool! Feel also free to tackle modifiers or members if you like, I haven't done anything there yet and for the rest of this week I'm working on other things.

**Comment by olafurpg** [comment]
@Dveim Do you have any un-merged work on Lambdas? I'm working on ValDef patterns in #114.

**Comment by Dveim** [comment]
Nothing important, merge won't be a problem.

I received bigger amount of work (at primary job) then expected. Sorry for that delay, I hope to get more free time soon.

**Comment by olafurpg** [comment]
No worries, just good to know. I'm working on the converter full-time this week and I've made a ton of progress. I may take lambdas then if that appears to be the next big blocker.

**Comment by Dveim** [comment]
Sure, if that blocks.

**Comment by olafurpg** [comment]
@xeno-by Could you please clarify members, modifiers and lambdas?

**Comment by olafurpg** [comment]
Quick status update. I did a small experiment to see which parts of m.Tree are uncovered by the converter:

```
Ctor.Ref.Function
Ctor.Ref.Name
Ctor.Ref.Project
Ctor.Ref.Select
Defn.Macro
Enumerator.Generator
Enumerator.Guard
Enumerator.Val
Mod.Inline
Pat.ExtractInfix
Pat.Type.And
Pat.Type.ApplyInfix
Pat.Type.Existential
Pat.Type.Function
Pat.Type.Or
Pat.Type.Placeholder
Pat.Type.Project
Pat.Type.Refine
Pat.Type.Tuple
Pat.Xml
Term.ApplyUnary
Term.For
Term.ForYield
Term.TryWithTerm
Term.Update
Term.Xml
Type.And
Type.Arg.Repeated
Type.Or
Type.Placeholder
```

Out of a corpus of ~26k source files (from scala/scala, intellij-scala, scala.js, akka, ...), the converter crashes on (only) ~120 source files, split between these errors

```
ConvertException: unexpected result: expected = Pat, actual = Pat.Type.Apply: 2
ConvertException: PackageDef: 2
ConvertException: unexpected result: expected = Pat.Type, actual = Type.Function: 3
ConvertException: Bind: 5
ConvertException: Ident: 5
ConvertException: ModuleDef: 11
ConvertException: unexpected result: expected = Pat.Type, actual = Type.Project: 12
UnreachableError: this code path should've been unreachable: 17
ConvertException: unexpected error (scroll down the stacktrace to see the cause):: 66
```

**Comment by xeno-by** [comment]
@olafurpg Amazing job! How about we replace the first comment in this thread with your latest comment?

**Comment by DavidDudson** [comment]
@olafurpg Could we update the main post again? I have a feeling some a few can be ticked off now

**Comment by olafurpg** [comment]
@DavidDudson I can't recall where I put the script to produce that report :( since there's so few left, I suspect it's easiest to grep for the remaining names in ToMtree to see if they're covered or not

**Comment by DavidDudson** [comment]
@olafurpg okay, we probably should move this issue to meta also, since we no longer keep the converter in paradise

**Comment by DavidDudson** [comment]
@xeno-by Can we close this please, given the new issue in meta

**Comment by olafurpg** [comment]
Thank you for moving the issue to the scalameta repo @DavidDudson! Closing this  one.
